src/dragonball/src/device_manager/blk_dev_mgr.rs

```diff
@@ -0,0 +1,769 @@
+        if block_device_config.is_root_device {
+            if self.has_root_block {
+                return Err(BlockDeviceError::RootBlockDeviceAlreadyAdded);
+            } else {
+                self.has_root_block = true;
+                self.read_only_root = block_device_config.is_read_only;
+                self.has_partuuid_root = block_device_config.partuuid.is_some();
+                self.partuuid = block_device_config.partuuid.clone();
+                // Root Device should be the first in the list whether or not PARTUUID is specified
+                // in order to avoid bugs in case of switching from partuuid boot scenarios to
+                // /dev/vda boot type.
+                self.info_list
+                    .push_front(BlockDeviceInfo::new(block_device_config));
+            }
+        } else {
+            self.info_list
+                .push_back(BlockDeviceInfo::new(block_device_config));
+            index = self.info_list.len() - 1;
+        }
+
+        Ok(index)
+    }
+
+    /// Updates a Block Device Config. The update fails if it would result in two
+    /// root block devices.
+    fn update(
+        &mut self,
+        mut index: usize,
+        new_config: BlockDeviceConfigInfo,
+    ) -> std::result::Result<(), BlockDeviceError> {
+        // Check if the path exists
+        self.check_data_file_present(&new_config)?;
+        if let Some(idx) = self.get_index_of_drive_path(&new_config.path_on_host) {
+            if idx != index {
+                return Err(BlockDeviceError::BlockDevicePathAlreadyExists(
+                    new_config.path_on_host.clone(),
+                ));
+            }
+        }
+
+        // Check if the root block device is being updated.
+        if self.info_list[index].config.is_root_device {
+            self.has_root_block = new_config.is_root_device;
+            self.read_only_root = new_config.is_root_device && new_config.is_read_only;
+            self.has_partuuid_root = new_config.partuuid.is_some();
+            self.partuuid = new_config.partuuid.clone();
+        } else if new_config.is_root_device {
+            // Check if a second root block device is being added.
+            if self.has_root_block {
+                return Err(BlockDeviceError::RootBlockDeviceAlreadyAdded);
+            } else {
+                // One of the non-root blocks is becoming root.
+                self.has_root_block = true;
+                self.read_only_root = new_config.is_read_only;
+                self.has_partuuid_root = new_config.partuuid.is_some();
+                self.partuuid = new_config.partuuid.clone();
+
+                // Make sure the root device is on the first position.
+                self.info_list.swap(0, index);
+                // Block config to be updated has moved to first position.
+                index = 0;
+            }
+        }
+        // Update the config.
+        self.info_list[index].config = new_config;
```

Comment:
Call it paranoia but I personally don't think we should ever index directly like this: it's too fragile if the surrounding code changes (correct today, but who knows about tomorrow...?)

tbh I'm kinda amazed rust even allows it given it's safety assurances as this just side-steps all that and crashes on invalid indices...

Reply:
index check added here.

Yes, I am also amazed that Rust doesn't have a safety check for this scenario. :o